runtime/minor_gc.c

```diff
@@ -683,18 +767,35 @@ void caml_empty_minor_heap_promote (struct domain* domain, int participating_cou
   atomic_store_rel((atomic_uintnat*)&domain_state->young_ptr, (uintnat)domain_state->young_end);
 
   if( not_alone ) {
-    atomic_fetch_add_explicit(&domains_finished_minor_gc, 1, memory_order_release);
+      while (1) {
+        finished_minor_gc = atomic_load_explicit(&domains_finished_minor_gc, memory_order_acquire);
+        if ((finished_minor_gc + 1) == participating_count) {
+          /* last domain to be done with minor collection, reset global allocation pointer */
+          caml_adjust_global_minor_heap(participating_count);
+        }
+
+        /* CAS away finished_minor_gc, only one domain should be the last one. */
+        if (atomic_compare_exchange_strong(&domains_finished_minor_gc,
+					 &finished_minor_gc,
+					 finished_minor_gc + 1)) {
+        	break;
+        }
+    }
+  }
+  else {
+    caml_adjust_global_minor_heap(1);
   }
 
   domain_state->stat_minor_words += Wsize_bsize (minor_allocated_bytes);
+
   domain_state->stat_minor_collections++;
   domain_state->stat_promoted_words += domain_state->allocated_words - prev_alloc_words;
 
   CAML_EV_END(EV_MINOR);
   caml_gc_log ("Minor collection of domain %d completed: %2.0f%% of %u KB live, rewrite: successes=%u failures=%u",
                domain->state->id,
-               100.0 * (double)st.live_bytes / (double)minor_allocated_bytes,
-               (unsigned)(minor_allocated_bytes + 512)/1024, rewrite_successes, rewrite_failures);
+               100.0 * (double)st.live_bytes / (double)(domain_state->stat_minor_words),
```

Comment:
I don't understand this change from `minor_allocated_bytes` to `stat_minor_words`.
The change makes the fraction in units `bytes`/`words`, rather than dimensionless; i.e. wouldn't we need to alter `st.live_bytes` to be _live words_.
The second change probably needs us to alter `KB live` to `K words live`.

Reply:
This change is to take into account that now, we may have used more minor heap space than just one segment (the one currently set to the domain.)
If the domain cycled through many minor heap segment with DLABs, we keep track of these in the replenish cycle: every time we are done doing replenish, we set `stat_minor_words` to `+=` the previous minor heap segment.
This effectively allow us to keep track of the previous minor heap segments used by this domain.
However it means we do need to use `stat_minor_words` instead of the more local `minor_allocated_bytes` (that only knows of the current minor heap segment.).
I think a simpler fix would be to change `(domain_state->stat_minor_words)` to `Bsize_wsize(domain_state->stat_minor_words)` ?